apps/web/src/validators/pool/assessmentPlan.ts

```diff
@@ -3,13 +3,32 @@ import { parseDateTimeUtc } from "@gc-digital-talent/date-helpers";
 import { notEmpty } from "@gc-digital-talent/helpers";
 import {
   AssessmentStepType,
-  Pool,
+  FragmentType,
   PoolSkillType,
+  getFragment,
 } from "@gc-digital-talent/graphql";
 
 import { PoolCompleteness } from "~/types/pool";
+import ViewPool_Fragment from "~/pages/Pools/ViewPoolPage/fragment";
+import AssessmentPlanBuilderPool_Fragment from "~/pages/Pools/AssessmentPlanBuilderPage/fragment";
+
+function isViewPool_Fragment(
+  poolQuery:
+    | FragmentType<typeof ViewPool_Fragment>
+    | FragmentType<typeof AssessmentPlanBuilderPool_Fragment>,
+): poolQuery is FragmentType<typeof ViewPool_Fragment> {
+  return "ViewPoolFragment" in poolQuery;
+}
+
+export function getAssessmentPlanStatus(
+  poolQuery:
+    | FragmentType<typeof ViewPool_Fragment>
+    | FragmentType<typeof AssessmentPlanBuilderPool_Fragment>,
+): PoolCompleteness {
+  const pool = isViewPool_Fragment(poolQuery)
+    ? getFragment(ViewPool_Fragment, poolQuery)
+    : getFragment(AssessmentPlanBuilderPool_Fragment, poolQuery);
```

Comment:
So, while this works I beleive we decided to be more explicit about specific items we pass in rather than pass around fragments or full model types. So I'd probably change this function signature to look more like:


```tsx
type GetAssessmentPlanStatusArgs = {
   poolSkills: Maybe<Array<Maybe<Pick<PoolSkill, "id" | "type">>> | undefined;
   addessmentSteps: Maybe<Array<Maybe<Pick<AssessmentStep, "step">>> | undefined;
}

type GetAssessmentPlanStatusFunc = (args: GetAssessmentPlanStatusArgs): PoolCompleteness;
```

Or if that doesn't work, you could just define a fragment for this function:

```tsx
const AssessmentPlanStatus_Fragment = graphql(/* GraphQL */`
    fragment AssessmentPlanStatus on Pool {
        poolSkills {
          id
          type
        }
        assessmentSteps {
          step {
             poolSkills { id }
          }
        }
    }
`);

function getAssessmentPlanStatus(poolQuery: FragmentType<AssessmentPlanStatus_Fragment>) {
  const pool = getFragment(AssessmentPlanStatus_Fragment, poolQuery);
}
```

Reply:
Second recommendation does not seem to work for me as it complains about a fragment type mismatch

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/44a772ab-8720-454d-8e86-51c671aa4d5f)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/3a99445b-20d2-4d75-a4eb-5f46d0262e7c)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/dbf2f1cc-2d54-4351-b499-f2745b93bd9f)

Reply:
First recommendation also doesn't seem to work for me 😕 

If I define the function like so with nullable `publishedAt`
I get this error

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/6c98c4cb-103d-4a01-8e1f-e90675e7238f)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/6f585923-e6cd-407b-a020-3c317bbc2f75)

If I define it so it isn't nullable
I get this error instead

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/b7566dc2-46ab-4ee5-b5ff-c62fbc8b28df)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/b8be79d4-9e45-4335-b851-fd7da658d2e0)

Reply:
For the first one, I tihnk you may need to pass it in like `getAssessmentPlanStatus(pool);` and not the fragment itself. I assume that is where the issue is coming from.

On the second one, I think the type for `publishedAt` is just a tiny bit off. Could you try: `pulbishedAt?: Maybe<string>;`?

Reply:
Adding the question marks stopped errors. But now it does not warn about missing fields, like assessmentSteps not being queried at all.
I do not think mixing typescript typing with fragments is fruitful

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/e6b18560-c615-4a52-b0f4-ff326b9ecf71)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/1271cced-aa53-42ba-9d01-4663fbca464c)

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/6e05d1f0-a4a4-4630-bdca-2d533259b508)

Reply:
> For the first one, I tihnk you may need to pass it in like `getAssessmentPlanStatus(pool);` and not the fragment itself. I assume that is where the issue is coming from.
> 
> On the second one, I think the type for `publishedAt` is just a tiny bit off. Could you try: `pulbishedAt?: Maybe<string>;`?

That doesn't seem to work either 😢 

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/205a7299-83ed-47c8-9e67-53ab1c13ede6)

Reply:
Yeah, I think we should be refactoring this to not pass the entire pool in to the function. We should only be passing in what the function needs to work. So something like:

```tsx
const assessmentStatus = getAssessmentStatus(
  pool.publishedAt,
  unpackMaybes(pool.poolSkills),
  unpackMaybes(pool.assessmentSteps)
);
```

That way we see the error here when passing items in and it requires them to exist. Now, this does not solve the issue with sub-fields. But, we could take it a step further and be very explicit about our types :thinking: but that seems overly verbose :sweat_smile: 

You will see, I removed the `assessmentSteps` entirely and it was able to pick up the error with the basic type. Then, if I removed the `PoolSkill.type` from the fragment and typed it explicitly, it was able to warn about a sub field missing.

![screenshot_02-May-2024_15-45-1714679140](https://github.com/GCTC-NTGC/gc-digital-talent/assets/4127998/6072ec4c-e508-4093-abe3-5e3017e72798)